File: geeksgforgeeks/challenge11_cuttingarod/CuttingRod.java
/*
    Created by Avinash on 11th November, 2017
*/
import java.io.*;
import java.util.*;
public class CuttingRod
{
	public static int output;
	public static void main(String[] args)
	{
		BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
		int maxPrice = 0;
		try
		{
		System.out.println("Enter the length of rod : ");
		int rodLength;
        	rodLength = Integer.parseInt(read.readLine());
        	int[] price = new int[rodLength];
        	System.out.print("Enter the price of different lenghts of rod: ");
        	for(int i = 0; i < rodLength; i++)
        		price[i] = Integer.parseInt(read.readLine());
        	Helper link = new Helper();
        	maxPrice = link.findMaxPriceDP(price,rodLength);
		}
		catch(IOException e)
		{
			e.printStackTrace();
		}
		System.out.print("The maximum price that can be obtained is " +maxPrice);
	}
}

class Helper
{
        public int findMaxPrice(int[] price, int length)
	{
		if(length <= 0)
			return 0;
		int max = Integer.MIN_VALUE;
		for(int i = 0; i < length; i++)
			max = Math.max(max, price[i] + findMaxPrice(price, length - i - 1));
		return max;
	}
        
        public int findMaxPriceDP(int[] price, int length)
	{
		
		int cache[] = new int[length+1];
		int maxPrice = -1;;
		for(int i = 1; i <= length; i++)
		{
			int max = Integer.MIN_VALUE;
			for(int j = 0; j < i; j++)
			{
				maxPrice= Math.max(maxPrice, price[j] + cache[i - j - 1]);		
				cache[i] = maxPrice;
			 }
		 }
		 return cache[length];
	}
  	
}	                    	
			

File: geeksgforgeeks/challenge11_cuttingarod/cuttingrod.py
# Created by Avinash on 13th Nov, 2017
import sys
def process():
	n = int(input("Enter the length of the rod : "))
	price = [ 0 for i in range(n)]
	print("Enter the price of different lengths of rod : ")
	for i in range(n):
		price[i] = int(input())
	maxPrice = findMaxPriceDP(price,n)
	print("The maximum price that can be obtained is %-5d" %maxPrice)

def findMaxPrice(price,length):
	if length <= 0 :
		return 0
	maxPrice = -sys.maxsize
	for i in range(0,length):
		maxPrice = max(maxPrice, price[i] + findMaxPrice(price,length-i-1))
	return maxPrice

def findMaxPriceDP(price, length):
	cache = [0 for i in range(0,length+1)]
	maxPrice = -1
	for i in range(1,length+1):
		for j in range(0,i):
			maxPrice = max(maxPrice,price[j]+cache[i-j-1])
			cache[i] = maxPrice
	return cache[length]	

				

if __name__=="__main__":
	process()
		
			